core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -813,7 +813,17 @@ class Partition(val topicPartition: TopicPartition,
    */
   private def tryCompleteDelayedRequests(): Unit = delayedOperations.checkAndCompleteAll()
 
-  def maybeShrinkIsr(): Unit = {
+  private def shouldShrinkIsr(): Boolean = {
+    leaderLogIfLocal match {
+      case Some(leaderLog) =>
+        inReadLock(leaderIsrUpdateLock) {
+          getOutOfSyncReplicas(replicaLagTimeMaxMs).nonEmpty
+        }
+      case None => false
+    }
+  }
+
+  private def tryShrinkIsr(): Unit = {
```

Comment:
This change looks good to me overall. My only question is whether we need both tryShrinkIsr and maybeShrinkIsr, and whether we could rename tryShrinkIsr to maybeShrinkIsr and perform the shouldShrinkIsr check within it.

I think we should also add a comment explaining the behavior so we don't see regressions in the future.

Reply:
Merged both - done.